README.md
# homebridge-nhc2
[![Version](http://img.shields.io/npm/v/@openhomekit/homebridge-nhc2.png)](https://www.npmjs.org/package/@openhomekit/homebridge-nhc2)
[![License](https://img.shields.io/npm/l/@openhomekit/homebridge-nhc2.svg)](https://github.com/homebridge-nhc2/homebridge-nhc2/blob/master/LICENSE)
[![Downloads](https://img.shields.io/npm/dt/@openhomekit/homebridge-nhc2.svg)](https://www.npmjs.org/package/@openhomekit/homebridge-nhc2)
[![Node.js Package](https://github.com/openhomekit/homebridge-nhc2/actions/workflows/npm-publish.yml/badge.svg)](https://github.com/openhomekit/homebridge-nhc2/actions/workflows/npm-publish.yml)

As of October 29, 2019 Niko has published their [Hobby API](https://www.niko.eu/en/campaign/niko-home-control/hobby-api) which allows end users to control their Niko Home Control 2 installation through the MQTT protocol.
This homebridge plugin adds homekit support for a Niko Home Control 2 having the Hobby API defined as a connected service.

Supported actions include:
* toggle light status
* set light brightness level

## Changes

You can read the complete history of changes in the 
[CHANGELOG](https://github.com/homebridge-nhc2/homebridge-nhc2/blob/master/CHANGELOG.md).

## Known Issues

1. Currently we need to set the `rejectUnauthorized` option to false because we are unable to verify the Niko root CA which is not ideal. Feel free to look into this.

## Project Principles

This project has a few principles that have and will continue to guide its 
development.

1. **Dependency lean**. Try to keep the required dependencies to a minimum.
2. **Simple**. Using the plugin should be simple and straightforward 
following common conventions.
3. **Completeness** This plugin is a far way from being complete, but we aim to make this plugin feature complete based on the official Niko documentation.

## Contributing

Contributions are welcome, particularly bug fixes and enhancements!
Refer to our [Contribution Guidelines](https://github.com/homebridge-nhc2/homebridge-nhc2/blob/master/CONTRIBUTING.md) for details.

> Please note that Project owners reserve the right to accept or reject any PR
> for any reason.

## Code of Conduct

Before contributing or participating in the homebridge-nhc2 community please be sure to 
familiarize yourself with our project 
[CODE OF CONDUCT](https://github.com/homebridge-nhc2/homebridge-nhc2/blob/master/CODE_OF_CONDUCT.md). 
These guidelines are intended to govern interactions with and within the homebridge-nhc2 
community.

# Hobby API Documentation

The Hobby API encapusulated by this plugin is documented by Niko which can be found [here](https://mynikohomecontrol.niko.eu/Content/hobbyapi/documentation.pdf)
	
# Warranty Disclaimer

You may use this plugin with the understanding that doing so is 
**AT YOUR OWN RISK**. No warranty, express or implied, is made with regards 
to the fitness or safety of this code for any purpose. If you use this 
plugin to query or change settings of your installation you understand that it 
is possible to break your installation and may require the replace of devices or 
intervention of professionals of which costs cannot be returned by the project team.

# Installation

In order to use the plugin you must first download and install it globally.

    npm install -g @openhomekit/homebridge-nhc2

You may also install directly from the GitHub 
[source](https://github.com/homebridge-nhc2/homebridge-nhc2). Either download and unzip 
the source, or clone the repository. Run the build command and pass the **-P** flag to homebridge to be able to locate the plugin.

## Connecting the Hobby API

Login in to [mynikohomecontrol](https://mynikohomecontrol.niko.eu/) and add the `Hobby API` to your connected services. This will provide you with a password which in formatted as a JWT token and valid for 1 year.

Following options are available for configuring the plugin.
* host 
* port (default to 8884)
* clientId (default to NHC2-homebridge)
* username (default to hobby)
* password

	  
Add the plugin to the **platforms** section in your homebridge configuration file.
```json
  "platforms": [
    {
      "platform" : "NHC2",
      "name" : "NHC2",
      "host": "<IP_ADDRESS_OF_YOUR_CONNECTED_CONTROLLER>",
      "password": "<PASSWORD_PROVIDED_BY_MYNIKOHOMECONTROLL>"
    }
  ]
```

> Note: The clientId should be unique to the MQTT service provided by the Connected Controller. 
> If multiple connections with the same ClientId are running these will continuously 
> disconnect/reconnect and may skip MQTT messages.

### More Configuration Options

#### Verbose Logging

Another configuration option is de `verbose` switch. It cranks up the information that is published to the logs. This is still at an informational level and can help in detecting anomalies. To really see debugging level details, run homebridge with the `-D` switch.

#### Suppressing Accessories

Besides the required `host` & `password` and other connection-related configuraion options, the configuration also allows for suppressing accessories:

```json
  "platforms": [
    {
      "platform" : "NHC2",
      "name" : "NHC2",
      "host": "<IP_ADDRESS_OF_YOUR_CONNECTED_CONTROLLER>",
      "password": "<PASSWORD_PROVIDED_BY_MYNIKOHOMECONTROLL>",
      "suppressedAccessories": [
          "fa33d687-9225-4f9e-b55e-013abb69b42e"
      ]
    }
  ]
```

This allows for not exposing certain accessories to Homebrigde (and therefore also the Home app). This feature can be useful when sharing access to the Home app with others. 

#### Inverting Operation of Windowblind

Interpreting the concept of open or closed is subjective to say the least. In case the interpretation of NHC and HomeKit doesn't match your personal preference (e.g. I like the concept of the slider to match the height of the rolled out blid) you might want to invert the operation of blinds. By adding the accessory's UUID to the configurable list of `invertedBlinds`, you can have it translated on the go:

```json
  "platforms": [
    {
      "platform" : "NHC2",
      "name" : "NHC2",
      "host": "<IP_ADDRESS_OF_YOUR_CONNECTED_CONTROLLER>",
      "password": "<PASSWORD_PROVIDED_BY_MYNIKOHOMECONTROLL>",
      "invertedBlinds": [
          "fa33d687-9225-4f9e-b55e-013abb69b42e"
      ]
    }
  ]
```

## Plugin Development

When working on this plugin, you'll want Homebridge to load it from your development directory instead of publishing it to `npm` each time.

Run this command inside your plugin project folder so your global install of Homebridge can discover it:


```shell
npm link
```

*You can undo this using the `npm unlink` command.*

Then start Homebridge in debug mode:

```shell
homebridge -D
```

This will start up Homebridge and load your in-development plugin. Note that you can also direct Homebridge to load your configuration from somewhere besides the default `~/.homebridge`, for example:

```shell
homebridge -D -U ./homebridge-dev
```

This is very useful when you are already using your development machine to host a "real" Homebridge instance (with all your accessories) that you don't want to disturb.

# Releasing a new version
To release a new version it is sufficient to bump the version number and push the changes, then create a new version on github that will create a tag for you. Github CI will see the new version and publish to NPM.

Please make sure your released version is supported by [HOOBS](https://hoobs.org/), you can run HOOBS as a docker container for testing purposes.

```docker run -p 8080:80 hoobs/hoobs```


CHANGELOG.md
# Changelog

## v2.0.0
* Added support for suppressing accessories
  
* Added actions
  * toggle basic state change on generic devices (free start stop actions)
  * change position (e.g. of sunblinds)
  * toggle light status
  * set light brightness level

## v3.1.0

* Filtered out devices that are not controlable (only actions and routines)

* Added actions
  * Basic fan control (switched-fan)
  * Ventilation control (fan)
    * 0 = Low 
    * 1-49 = Medium 
    * 50-99 = High 
    * 100 = Boost
* Added routines
  * All-Off
  * Presence Simulation
  * Alarms
  * Scenes (comfort)
  * Peak Mode
  * Solar Mode

## v3.2.0

* Added support for thermostats

## v3.3.0

* Added support for thermostat programs
  * Homebridge --> Niko
    * Off       -->   Off
    * Heating   -->   Day
    * Cooling   -->   Cool
    * Auto      -->   Prog 1
  * Any other programs selected in Niko will display as Auto
  * Min temp for homebridge is 10 where as Niko supports as low as 7. 
    * Any temp set under 10 degrees in Niko will be displayed as 10. 
  * Changed the default override time from 24h to 60min

src/nhc2-platform.ts
import { Demand } from "@openhomekit/nhc2-hobby-api/lib/event/Demand";
import { Device } from "@openhomekit/nhc2-hobby-api/lib/event/device";
import { Event } from "@openhomekit/nhc2-hobby-api/lib/event/event";
import { FanSpeed } from "@openhomekit/nhc2-hobby-api/lib/event/FanSpeed";
import { Program } from "@openhomekit/nhc2-hobby-api/lib/event/Program";
import { NHC2 } from "@openhomekit/nhc2-hobby-api/lib/NHC2";
import {
  API,
  APIEvent,
  Characteristic,
  CharacteristicEventTypes,
  CharacteristicSetCallback,
  CharacteristicValue,
  DynamicPlatformPlugin,
  HAP,
  Logging,
  PlatformAccessory,
  PlatformConfig,
  Service,
} from "homebridge";

import { NHC2Logger } from "./nhc2-logger";

const PLUGIN_NAME = "homebridge-nhc2";
const PLATFORM_NAME = "NHC2";

let hap: HAP;
let Accessory: typeof PlatformAccessory;

export = (api: API) => {
  hap = api.hap;
  Accessory = api.platformAccessory;

  api.registerPlatform(PLUGIN_NAME, PLATFORM_NAME, NHC2Platform);
};

class NHC2Platform implements DynamicPlatformPlugin {
  private readonly Service: typeof Service = this.api.hap.Service;
  private readonly Characteristic: typeof Characteristic = this.api.hap
    .Characteristic;

  private readonly accessories: PlatformAccessory[] = [];
  private readonly suppressedAccessories: string[] = [];
  private readonly nhc2: NHC2;

  private readonly log: NHC2Logger;

  constructor(
    private logger: Logging,
    private config: PlatformConfig,
    private api: API,
  ) {
    this.log = new NHC2Logger(this.logger, this.config);
    this.suppressedAccessories = config.suppressedAccessories || [];
    if (this.suppressedAccessories) {
      this.log.info("Suppressing accessories: ");
      this.suppressedAccessories.forEach(acc => {
        this.log.info("  - " + acc);
      });
    }
    this.nhc2 = new NHC2("mqtts://" + this.config.host, {
      port: this.config.port || 8884,
      clientId: this.config.clientId || "NHC2-homebridge",
      username: this.config.username || "hobby",
      password: this.config.password,
      rejectUnauthorized: false,
    });

    this.log.info("NHC2Platform finished initializing!");

    api.on(APIEvent.DID_FINISH_LAUNCHING, async () => {
      this.log.info("NHC2Platform 'didFinishLaunching'");

      await this.nhc2.subscribe();
      const nhc2Accessories = await this.nhc2.getAccessories();
      this.log.info("got " + nhc2Accessories.length + " accessories");
      this.addAccessories(nhc2Accessories);

      this.nhc2.getEvents().subscribe(event => {
        this.processEvent(event);
      });
    });
  }

  public configureAccessory(accessory: PlatformAccessory): void {
    this.accessories.push(accessory);
  }

  public processEvent = (event: Event) => {
    this.log.debug('Event: ', JSON.stringify(event))
    if (!!event.Params) {
      event.Params.flatMap(param =>
        param.Devices.forEach((device: Device) => {
          const deviceAccessoryForEvent = this.findAccessoryDevice(device);
          if (!!deviceAccessoryForEvent) {
            deviceAccessoryForEvent.services.forEach(service =>
              this.processDeviceProperties(device, service),
            );
          }
        }),
      );
    }
  };

  private findAccessoryDevice(device: Device) {
    return this.accessories.find(accessory => accessory.UUID === device.Uuid);
  }

  private addAccessories(accessories: Device[]) {
    const mapping: { [index: string]: any } = {
      light: {
        service: this.Service.Lightbulb,
        handlers: [this.addStatusChangeCharacteristic]
      },
      dimmer: {
        service: this.Service.Lightbulb,
        handlers: [
          this.addStatusChangeCharacteristic,
          this.addBrightnessChangeCharacteristic
        ],
      },
      socket: {
        service: this.Service.Outlet,
        handlers: [this.addStatusChangeCharacteristic]
      },
      generic: {
        service: this.Service.Switch,
        handlers: [this.addTriggerCharacteristic]
      },
      'switched-generic': {
        service: this.Service.Switch,
        handlers: [this.addStatusChangeCharacteristic]
      },
      'switched-fan': {
        service: this.Service.Fan,
        handlers: [this.addStatusChangeCharacteristic]
      },
      sunblind: {
        service: this.Service.WindowCovering,
        handlers: [this.addPositionChangeCharacteristic]
      },
      venetianblind: {
        service: this.Service.WindowCovering,
        handlers: [this.addPositionChangeCharacteristic]
      },
      rolldownshutter: {
        service: this.Service.WindowCovering,
        handlers: [this.addPositionChangeCharacteristic]
      },
      gate: {
        service: this.Service.WindowCovering,
        handlers: [this.addPositionChangeCharacteristic]
      },
      alloff: {
        service: this.Service.Switch,
        handlers: [this.addTriggerCharacteristic]
      },
      simulation: {
        service: this.Service.Switch,
        handlers: [this.addTriggerCharacteristic]
      },
      alarms: {
        service: this.Service.Switch,
        handlers: [this.addTriggerCharacteristic]
      },
      comfort: {
        service: this.Service.Switch,
        handlers: [this.addStatusChangeCharacteristic]
      },
      peakmode: {
        service: this.Service.Switch,
        handlers: [this.addTriggerCharacteristic]
      },
      solarmode: {
        service: this.Service.Switch,
        handlers: [this.addTriggerCharacteristic]
      },
      fan: {
        service: this.Service.Fan,
        handlers: [
          this.addOnFanCharacteristic,
          this.addOffFanCharacteristic
        ]
      },
      thermostat: {
        service: this.Service.Thermostat,
        handlers: [
          this.addTargetTemperatureCharacteristic,
          this.addProgramCharacteristic
        ]
      }
    };

    Object.keys(mapping).forEach(model => {
      const config = mapping[model];
      const accs = accessories.filter(
        acc =>
          !this.suppressedAccessories.includes(acc.Uuid)
          && acc.Model === model
          && (
            acc.Type === 'action' ||
            acc.Type === 'thermostat')
      );
      accs.forEach(acc => {
        const newAccessory = new Accessory(acc.Name as string, acc.Uuid);
        const newService = new config.service(acc.Name);
        config.handlers.forEach((handler: any) => {
          handler(newService, newAccessory);
        });
        newAccessory.addService(newService);
        this.processDeviceProperties(acc, newService);
        this.registerAccessory(newAccessory);
      });
    });
  }

  private registerAccessory(accessory: PlatformAccessory) {
    const existingAccessory = this.findExistingAccessory(accessory);
    if (!!existingAccessory) {
      this.unregisterAccessory(existingAccessory);
    }

    this.accessories.push(accessory);
    this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [
      accessory,
    ]);
    this.log.debug(
      "registered accessory: " +
      accessory.displayName +
      " (" +
      accessory.UUID +
      ")",
    );
  }

  private unregisterAccessory(accessory: PlatformAccessory) {
    this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [
      accessory,
    ]);
    this.accessories.splice(this.accessories.indexOf(accessory), 1);
    this.log.debug(
      "unregistered accessory: " +
      accessory.displayName +
      " (" +
      accessory.UUID +
      ")",
    );
  }

  private findExistingAccessory(newAccessory: PlatformAccessory) {
    return this.accessories
      .filter(accessory => accessory.UUID === newAccessory.UUID)
      .find(() => true);
  }

  private addStatusChangeCharacteristic = (
    newService: Service,
    newAccessory: PlatformAccessory,
  ) => {
    newService
      .getCharacteristic(this.Characteristic.On)
      .on(
        CharacteristicEventTypes.SET,
        (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
          this.nhc2.sendStatusChangeCommand(
            newAccessory.UUID,
            value as boolean,
          );
          callback();
        },
      );
  };

  private addBrightnessChangeCharacteristic = (
    newService: Service,
    newAccessory: PlatformAccessory,
  ) => {
    newService
      .getCharacteristic(this.Characteristic.Brightness)
      .on(
        CharacteristicEventTypes.SET,
        (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
          this.nhc2.sendBrightnessChangeCommand(
            newAccessory.UUID,
            value as number,
          );
          callback();
        },
      );
  };

  private addTriggerCharacteristic = (
    newService: Service,
    newAccessory: PlatformAccessory,
  ) => {
    newService
      .getCharacteristic(this.Characteristic.On)
      .on(
        CharacteristicEventTypes.SET,
        (_: CharacteristicValue, callback: CharacteristicSetCallback) => {
          this.nhc2.sendTriggerBasicStateCommand(newAccessory.UUID);
          callback();
        },
      );
  };

  private addOnFanCharacteristic = (
    newService: Service,
    newAccessory: PlatformAccessory,
  ) => {
    newService
      .getCharacteristic(this.Characteristic.RotationSpeed)
      .on(
        CharacteristicEventTypes.SET,
        (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
          switch (true) {
            case value === 100:
              this.nhc2.sendFanSpeedCommand(
                newAccessory.UUID,
                FanSpeed.Boost,
              );
              break;
            case value >= 50:
              this.nhc2.sendFanSpeedCommand(
                newAccessory.UUID,
                FanSpeed.High,
              );
              break;
            case value === 0:
              this.nhc2.sendFanSpeedCommand(
                newAccessory.UUID,
                FanSpeed.Low,
              );
              break;
            default:
              this.nhc2.sendFanSpeedCommand(
                newAccessory.UUID,
                FanSpeed.Medium,
              );
          }
          callback();
        }
      )
  };

  private addOffFanCharacteristic = (
    newService: Service,
    newAccessory: PlatformAccessory,
  ) => {
    newService
      .getCharacteristic(this.Characteristic.On)
      .on(
        CharacteristicEventTypes.SET,
        (_: CharacteristicValue, callback: CharacteristicSetCallback) => {
          this.nhc2.sendFanSpeedCommand(
            newAccessory.UUID,
            FanSpeed.Low,
          );
          callback();
        }
      )
  };

  private addPositionChangeCharacteristic =
    (
      newService: Service,
      newAccessory: PlatformAccessory
    ) => {
      newService
        .getCharacteristic(this.Characteristic.TargetPosition)
        .on(
          CharacteristicEventTypes.SET,
          (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
            this.nhc2.sendPositionChangeCommand(
              newAccessory.UUID,
              value as number,
            );
            callback();
          },
        );
    };

  private addTargetTemperatureCharacteristic =
    (
      newService: Service,
      newAccessory: PlatformAccessory
    ) => {
      newService
        .getCharacteristic(this.Characteristic.TargetTemperature)
        .on(
          CharacteristicEventTypes.SET,
          (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
            if (value as number === newService.getCharacteristic(this.Characteristic.CurrentTemperature).value) {
              this.nhc2.sendTempOverruleCommand(newAccessory.UUID, false, value as number)
            } else {
              this.nhc2.sendTempOverruleCommand(newAccessory.UUID, true, value as number, 60)
            }
            callback();
          },
        );
    };

  private addProgramCharacteristic =
    (
      newService: Service,
      newAccessory: PlatformAccessory
    ) => {
      newService
        .getCharacteristic(this.Characteristic.TargetHeatingCoolingState)
        .on(
          CharacteristicEventTypes.SET,
          (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
            switch (value) {
              case 0:
                this.nhc2.sendProgramCommand(
                  newAccessory.UUID,
                  Program.Off,
                );
                break;
              case 1:
                this.nhc2.sendProgramCommand(
                  newAccessory.UUID,
                  Program.Day,
                );
                break;
              case 2:
                this.nhc2.sendProgramCommand(
                  newAccessory.UUID,
                  Program.Cool,
                );
                break;
              default:
                this.nhc2.sendProgramCommand(
                  newAccessory.UUID,
                  Program.Prog1,
                );
            }
            callback();
          }
        )
    };

  private processDeviceProperties(device: Device, service: Service) {

    // Super hacky, but for some reason every device has two services, one we added and another "AccessoryInformation".
    // We should not be modifying AccessoryInformation (This gives warnings);
    // If a better solution comes along please create a PR
    if (service.constructor.name === 'AccessoryInformation') {
      return
    };

    if (!!device.Properties) {
      device.Properties.forEach(property => {
        if (property.Status === "On" || property.BasicState === "On") {
          service.getCharacteristic(this.Characteristic.On).updateValue(true);
        }
        if (property.Status === "Off" || property.BasicState === "Off") {
          service.getCharacteristic(this.Characteristic.On).updateValue(false);
        }
        if (!!property.Brightness) {
          service
            .getCharacteristic(this.Characteristic.Brightness)
            .updateValue(property.Brightness);
        }
        if (!!property.FanSpeed) {
          switch (property.FanSpeed) {
            case FanSpeed.Boost:
              service.getCharacteristic(this.Characteristic.RotationSpeed).updateValue(100);
              service.getCharacteristic(this.Characteristic.On).updateValue(true);
              break;
            case FanSpeed.High:
              service.getCharacteristic(this.Characteristic.RotationSpeed).updateValue(66);
              service.getCharacteristic(this.Characteristic.On).updateValue(true);
              break;
            case FanSpeed.Medium:
              service.getCharacteristic(this.Characteristic.RotationSpeed).updateValue(33);
              service.getCharacteristic(this.Characteristic.On).updateValue(true);
              break;
            case FanSpeed.Low:
              service.getCharacteristic(this.Characteristic.RotationSpeed).updateValue(0);
              service.getCharacteristic(this.Characteristic.On).updateValue(false);
              break;
          }
        }
        if (!!property.Position) {
          const moving = device.Properties?.find(p => p.Moving)?.Moving === "True";
          service
            .getCharacteristic(this.Characteristic.CurrentPosition)
            .updateValue(parseInt(property.Position, 10));
          service
            .getCharacteristic(this.Characteristic.PositionState)
            .updateValue(moving ? 1 : 2);
          /* TODO: find a way to determine INCREASING=1 or DECREASING=0 */

          if (!moving) {
            service
              .getCharacteristic(this.Characteristic.TargetPosition)
              .updateValue(parseInt(property.Position, 10));
          }
        }
        if (!!property.AmbientTemperature) {
          service.getCharacteristic(this.Characteristic.CurrentTemperature)
            .updateValue(parseFloat(property.AmbientTemperature));
          service.getCharacteristic(this.Characteristic.TemperatureDisplayUnits)
            .updateValue(0);
        }
        if (!!property.SetpointTemperature) {
          switch (true) {
            // Niko has a min temp of 7 which homebridge cannot handle
            case parseFloat(property.SetpointTemperature) < 10:
              service.getCharacteristic(this.Characteristic.TargetTemperature).updateValue(10);
              break;
            default:
              service.getCharacteristic(this.Characteristic.TargetTemperature)
                .updateValue(parseFloat(property.SetpointTemperature));
              break;
          }
        }
        if (!!property.Program) {
          switch (property.Program) {
            case Program.Off:
              service.getCharacteristic(this.Characteristic.TargetHeatingCoolingState).updateValue(0);
              break;
            case Program.Day:
              service.getCharacteristic(this.Characteristic.TargetHeatingCoolingState).updateValue(1);
              break;
            case Program.Cool:
              service.getCharacteristic(this.Characteristic.TargetHeatingCoolingState).updateValue(2);
              break;
            default:
              service.getCharacteristic(this.Characteristic.TargetHeatingCoolingState).updateValue(3);
              break;
          }
        }
        if (!!property.Demand) {
          switch (property.Demand) {
            case Demand.None:
              service.getCharacteristic(this.Characteristic.CurrentHeatingCoolingState).updateValue(0);
              break;
            case Demand.Heating:
              service.getCharacteristic(this.Characteristic.CurrentHeatingCoolingState).updateValue(1);
              service.getCharacteristic(this.Characteristic.TargetHeatingCoolingState).updateValue(1);
              break;
            case Demand.Cooling:
              service.getCharacteristic(this.Characteristic.CurrentHeatingCoolingState).updateValue(2);
              service.getCharacteristic(this.Characteristic.TargetHeatingCoolingState).updateValue(2);
              break;
          }
        }
      });
    }
  }
}
